crates/sui-types/src/dynamic_field.rs

```diff
@@ -206,28 +208,58 @@ pub fn is_dynamic_object(move_struct: &MoveStruct) -> bool {
 pub fn derive_dynamic_field_id<T>(
     parent: T,
     key_type_tag: &TypeTag,
-    key_type_layout: &MoveTypeLayout,
-    key: &move_vm_types::values::Value,
-) -> Option<ObjectID>
+    key_bytes: &[u8],
+) -> Result<ObjectID, bcs::Error>
 where
     T: Into<SuiAddress>,
 {
-    let Ok(k_tag_bytes) = bcs::to_bytes(key_type_tag) else {
-        return None;
-    };
-    let Some(k_bytes) = key.simple_serialize(key_type_layout) else {
-        return None;
-    };
+    let k_tag_bytes = bcs::to_bytes(key_type_tag)?;
 
     // hash(parent || key || key_type_tag)
     let mut hasher = Sha3_256::default();
     hasher.update(parent.into());
-    hasher.update(k_bytes.len().to_le_bytes());
-    hasher.update(k_bytes);
+    hasher.update(key_bytes.len().to_le_bytes());
+    hasher.update(key_bytes);
     hasher.update(k_tag_bytes);
     let hash = hasher.finalize();
 
     // truncate into an ObjectID and return
     // OK to access slice because Sha3_256 should never be shorter than ObjectID::LENGTH.
-    Some(ObjectID::try_from(&hash.as_ref()[0..ObjectID::LENGTH]).unwrap())
+    Ok(ObjectID::try_from(&hash.as_ref()[0..ObjectID::LENGTH]).unwrap())
+}
+
+pub trait MoveTypeTagTrait {
+    fn get_type_tag() -> TypeTag;
+}
```

Comment:
Consider moving this trait into `sui-types/src/lib.rs`? It seems more broadly relevant than just for dynamic fields, also, nit: consider dropping the `Trait` from the name?